// @soroban-react/chains/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.standalone = exports.sandbox = exports.testnet = exports.futurenet = exports.public_chain = void 0;
const SorobanClient = __importStar(require("soroban-client"));
/**
 * A `WalletChain` object representing the public blockchain network.
 * @property {string} id - The unique identifier for the blockchain network.
 * @property {string} name - The name of the blockchain network.
 * @property {string} networkPassphrase - The network passphrase for the blockchain network.
 */
exports.public_chain = {
    id: 'public',
    name: 'Public',
    networkPassphrase: SorobanClient.Networks.PUBLIC,
};
/**
 * A `WalletChain` object representing the Futurenet blockchain network.
 */
exports.futurenet = {
    id: 'public',
    name: 'Futurenet',
    networkPassphrase: "Test SDF Future Network ; October 2022",
};
/**
 * A `WalletChain` object representing the Testnet blockchain network.
 */
exports.testnet = {
    id: 'public',
    name: 'Testnet',
    networkPassphrase: SorobanClient.Networks.TESTNET,
};
/**
 * A `WalletChain` object representing the Sandbox blockchain network.
 */
exports.sandbox = {
    id: 'public',
    name: 'Sandbox',
    networkPassphrase: "Local Sandbox Stellar Network ; September 2022",
};
/**
 * A `WalletChain` object representing the Standalone blockchain network.
 */
exports.standalone = {
    id: 'public',
    name: 'Standalone',
    networkPassphrase: 'Standalone Network ; February 2017',
};
