// @soroban-react/contracts/dist/useContractValue.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useContractValue = void 0;
const react_1 = __importDefault(require("react"));
const SorobanClient = __importStar(require("soroban-client"));
let xdr = SorobanClient.xdr;
// Dummy source account for simulation. The public key for this is all 0-bytes.
const defaultAddress = 'GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAWHF';
// useContractValue is a hook that fetches the value of a contract method. It
// might be better named `useSimulateTransaction`, but not sure which is more clear...
// TODO: Allow user to specify the wallet of the submitter, fees, etc... Maybe
// a separate (lower-level) hook for `useSimulateTransaction` would be cleaner?
function useContractValue({ contractId, method, params, source, sorobanContext, }) {
    const { activeChain, address, server } = sorobanContext;
    const [value, setValue] = react_1.default.useState({ loading: true });
    const [xdrParams, setXdrParams] = react_1.default.useState(params ? params.map(p => p.toXDR().toString('base64')) : undefined);
    react_1.default.useEffect(() => {
        source = source !== null && source !== void 0 ? source : new SorobanClient.Account(address !== null && address !== void 0 ? address : defaultAddress, '0');
        if (!activeChain) {
            setValue({ error: 'No active chain' });
            return;
        }
        if (!server) {
            setValue({ error: 'Not connected to server' });
            return;
        }
        ;
        (() => __awaiter(this, void 0, void 0, function* () {
            setValue({ loading: true });
            try {
                let result = yield fetchContractValue({
                    server: server,
                    networkPassphrase: activeChain.networkPassphrase,
                    contractId: contractId,
                    method: method,
                    params: params,
                    source: source,
                });
                setValue({ result });
            }
            catch (error) {
                if (typeof error == 'string') {
                    setValue({ error });
                    return;
                }
                if ('message' in error) {
                    setValue({ error: error.message });
                    return;
                }
                setValue({ error });
            }
        }))();
        // Have this re-fetch if the contractId/method/params change. Total hack with
        // xdr-base64 to enforce real equality instead of object equality
        // shenanigans.
    }, [contractId, method, xdrParams, activeChain, server]);
    return value;
}
exports.useContractValue = useContractValue;
function fetchContractValue({ server, networkPassphrase, contractId, method, params, source, }) {
    return __awaiter(this, void 0, void 0, function* () {
        const contract = new SorobanClient.Contract(contractId);
        let myParams = params || [];
        // TODO: Optionally include the wallet of the submitter here, so the
        // simulation is more accurate
        const transaction = new SorobanClient.TransactionBuilder(source, {
            // fee doesn't matter, we're not submitting
            fee: '100',
            networkPassphrase,
        })
            .addOperation(contract.call(method, ...myParams))
            .setTimeout(SorobanClient.TimeoutInfinite)
            .build();
        const { results } = yield server.simulateTransaction(transaction);
        if (!results || results.length !== 1) {
            throw new Error('Invalid response from simulateTransaction');
        }
        const result = results[0];
        return xdr.ScVal.fromXDR(result.xdr, 'base64');
    });
}
